src/partition/ReceivedPartition.cpp

```diff
@@ -33,30 +36,61 @@ ReceivedPartition::ReceivedPartition(
 void ReceivedPartition::communicate()
 {
   PRECICE_TRACE();
-  PRECICE_INFO("Receive global mesh " << _mesh->getName());
-  Event e("partition.receiveGlobalMesh." + _mesh->getName(), precice::syncMode);
-  if (not utils::MasterSlave::isSlave()) {
-    PRECICE_ASSERT(_mesh->vertices().empty());
-    // a ReceivedPartition can only have one communication, @todo nicer design
-    com::CommunicateMesh(_m2ns[0]->getMasterCommunication()).receiveMesh(*_mesh, 0);
+  PRECICE_ASSERT(_m2ns.size() == 1);
+  PRECICE_ASSERT(_mesh->vertices().empty());
+
+  if (_m2ns[0]->usesTwoLevelInitialization()) {
+    PRECICE_INFO("Receive mesh partitions for mesh " << _mesh->getName());
+    Event e("partition.receiveMeshPartitions." + _mesh->getName(), precice::syncMode);
+
+    if (utils::MasterSlave::isMaster()) {
+      // Master receives remote mesh's global number of vertices
+      int globalNumberOfVertices = -1;
+      _m2ns[0]->getMasterCommunication()->receive(globalNumberOfVertices, 0);
+      _mesh->setGlobalNumberOfVertices(globalNumberOfVertices);
+    }
+
+    // each rank receives max/min global vertex indices from connected remote ranks
+    _m2ns[0]->broadcastReceiveAll(_remoteMinGlobalVertexIDs, *_mesh);
+    _m2ns[0]->broadcastReceiveAll(_remoteMaxGlobalVertexIDs, *_mesh);
+    // each rank receives mesh partition from connected remote ranks
+    _m2ns[0]->broadcastReceiveAllMesh(*_mesh);
+
+  } else {
+    PRECICE_INFO("Receive global mesh " << _mesh->getName());
+    Event e("partition.receiveGlobalMesh." + _mesh->getName(), precice::syncMode);
+
+    if (not utils::MasterSlave::isSlave()) {
+      // a ReceivedPartition can only have one communication, @todo nicer design
+      com::CommunicateMesh(_m2ns[0]->getMasterCommunication()).receiveMesh(*_mesh, 0);
+      _mesh->setGlobalNumberOfVertices(_mesh->vertices().size());
+    }
+  }
+  if (utils::MasterSlave::isMaster()) {
```

Comment:
If this is a separate logical block, it may be good to explain its purpose in a comment.
Otherwise we could merge it with the above.